llms/mlx_lm/server.py

```diff
@@ -289,21 +313,42 @@ def handle_completion(
         finish_reason = "length"
         stop_sequence_suffix = None
         logging.debug(f"Starting completion:")
-        for (token, _), _ in zip(
+        log_probabilities = []
+        top_tokens = []
+        for (token, probability, step_logits), _ in zip(
             generate_step(
                 prompt=prompt,
                 model=self.model,
                 temp=self.temperature,
                 top_p=self.top_p,
                 repetition_penalty=self.repetition_penalty,
                 repetition_context_size=self.repetition_context_size,
+                return_step_logits=True,
                 logit_bias=self.logit_bias,
             ),
             range(self.max_tokens),
         ):
             detokenizer.add_token(token)
             logging.debug(detokenizer.text)
             tokens.append(token)
+
+            log_probs = mx.log(mx.softmax(step_logits, axis=-1))
+
+            if self.logprobs > 0:
+                sorted_indices = mx.argpartition(-log_probs, kth=self.logprobs, axis=-1)
+                top_n_indices = sorted_indices[:, : self.logprobs]
+                top_n_log_probs = mx.take_along_axis(log_probs, top_n_indices, axis=1)
+
+                top_token_info = {
+                    int(t.item()): l_prob.item()
```

Comment:
This cast to int is redundant but added because it seems the current server is serializing the tokens as integer strings despite the Python data dictionary having integers (the tokens as keys)